### ISSUE TITLE
Create Restricted Viewer

### ISSUE THREAD

**Issue by paul121** [Rothamsted-Ecoinformatics/farm_rothamsted#548]
This one has more complex permissions, specifically "View own" / "View associated with", that need to be tested in other places (dashboard, location hierarchy, etc).

General:
- [x] Create w/ update hook

Research perms:
- [x] Researcher: View own/Update own
- [x] Program: View own **
- [x] Proposal: View own **
- [x] Experiment: View own **
- [x] Plan: View own **
- [x] Cannot upload plot geoms
- [x] Cannot upload plot variables

Asset perms:

- [x] View any equipment
- [x] All others: View associated with 

Log perms:

- [x] All: View own, view associated with

Taxonomy:
- [x] View all

**Comment by paul121** [comment]
> This one has more complex permissions, specifically "View own" / "View associated with", that need to be tested in other places (dashboard, location hierarchy, etc).

I've done a bit of testing on this. I've added custom *access checks* for the "associated with" logic, but for these we *should* implement additional *query access checks/alters* to make filtering of entity pages work.

Fortunately, the Entity API module provides helpers for this, one example: https://gorannikolovski.com/blog/query-level-filtering-custom-entities-drupal-8

Our simplest use-case are Researcher entities where this approach works great:
- Researcher.farm_user -> Assigned to current user

With the `Restricted Viewer` role (specifically only having the `view assigned rothamsted_researcher` permission, not `view any rothamsted_researcher`), both Entity queries and the Researchers listing page is filtered to only show the one Researcher assigned to the user.

For other entities this gets more complicated due to the relationships required for the nested query:
- Program.principal_investigator -> Researcher -> Assigned to current user
- Proposal.contact -> Researcher -> Assigned to current user
- Plan -> Design -> Experiment -> Researcher -> Assigned to current user:
```
->condition('experiment_design.entity:rothamsted_design.experiment.entity:rothamsted_experiment.researcher.entity:rothamsted_researcher.farm_user', $account->id())
```
- This works fine for altering standard entity query access, but unfortunately the views query alter does not support the chained relationships. There is an existing issue *with a patch that does work!* but still needs further testing: https://www.drupal.org/project/entity/issues/3315432

For assets/logs this gets further complicated because we cannot do a single chained query, instead we need to do a `subquery` join/select *somewhere* because the relationship is: log-> asset, and plan -> asset, and plan -> ... as above.
- Existing issue demonstrates that subquery select is not currently possible with the Entity API solution, but likely possible with more custom work, using `group` module as an example: https://www.drupal.org/project/entity/issues/3259313

So TLDR; I have already implemented the *access checking* for this role - users cannot directly to go to the page of an entity they are not associated with. But unfortunately we cannot filter the lists of entities without some further work. I suggest we start by adding this role, but realize the UI will be poor.

**Comment by aislinnpearson** [comment]
Hi @paul121,

I thought this would be complicated... I've assigned it to myself as an issue for stress testing when you are ready to go with the demo site. 

Does it make sense to see how that goes, and then figure out any amendments (if needed)?

**Comment by paul121** [comment]
> Does it make sense to see how that goes, and then figure out any amendments (if needed)?

Yep I think so!
